File: ts/src/claudecode/mapping.ts
import type { Result, MessageType, Usage } from "../types.js";
import type { StreamMessage } from "./types.js";

/** Map Claude stream-json type to common MessageType. */
export function mapMessageType(type: string): MessageType {
  switch (type) {
    case "system":
      return "system";
    case "assistant":
      return "assistant";
    case "user":
      return "user";
    case "result":
      return "result";
    case "stream_event":
      // Stream events carry assistant content deltas.
      return "assistant";
    default:
      return type;
  }
}

/** Map a StreamMessage result to a common Result. */
export function mapResult(
  msg: StreamMessage,
  fallbackSessionId: string,
): Result {
  const usage: Usage = {
    inputTokens: msg.usage?.input_tokens ?? 0,
    outputTokens: msg.usage?.output_tokens ?? 0,
    cacheCreationInputTokens: msg.usage?.cache_creation_input_tokens ?? 0,
    cacheReadInputTokens: msg.usage?.cache_read_input_tokens ?? 0,
  };

  return {
    text: msg.result ?? "",
    isError: msg.is_error ?? false,
    exitCode: 0,
    usage,
    costUSD: msg.total_cost_usd ?? 0,
    durationMs: msg.duration_ms ?? 0,
    sessionId: msg.session_id || fallbackSessionId,
  };
}

File: python/src/agentrunner/claudecode/mapping.py
"""Map Claude stream-json types to common types."""

from __future__ import annotations

import json

from ..types import Result, Usage
from .types import StreamMessage

_TYPE_MAP = {
    "system": "system",
    "assistant": "assistant",
    "result": "result",
}


def map_message_type(raw_type: str, raw_line: str) -> str:
    """Map Claude stream-json type to common MessageType.

    Uses the raw JSON line to distinguish tool_use and tool_result subtypes
    from generic assistant/user messages, matching the INTERFACE.md taxonomy.
    """
    if raw_type in _TYPE_MAP:
        return _TYPE_MAP[raw_type]

    if raw_type == "user":
        try:
            d = json.loads(raw_line)
            for block in d.get("content", []):
                if isinstance(block, dict) and block.get("type") == "tool_result":
                    return "tool_result"
        except (json.JSONDecodeError, TypeError):
            pass
        return "user"

    if raw_type == "stream_event":
        return "assistant"

    return raw_type


def map_result(msg: StreamMessage, fallback_session_id: str) -> Result:
    """Map a StreamMessage result to a common Result."""
    usage = Usage(
        input_tokens=msg.usage.input_tokens if msg.usage else 0,
        output_tokens=msg.usage.output_tokens if msg.usage else 0,
        cache_creation_input_tokens=msg.usage.cache_creation_input_tokens if msg.usage else 0,
        cache_read_input_tokens=msg.usage.cache_read_input_tokens if msg.usage else 0,
    )

    return Result(
        text=msg.result or "",
        is_error=msg.is_error or False,
        exit_code=0,
        usage=usage,
        cost_usd=msg.total_cost_usd or 0,
        duration_ms=msg.duration_ms or 0,
        session_id=msg.session_id or fallback_session_id,
    )
